velox/functions/sparksql/ConcatWsFunctions.cpp

```diff
@@ -0,0 +1,202 @@
+      constantStringViews_.push_back(
+          StringView(constantString.data(), constantString.size()));
+    }
+  }
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& /* outputType */,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    context.ensureWritable(rows, VARCHAR(), result);
+    auto flatResult = result->asFlatVector<StringView>();
+    if (nullDelim_) {
+      flatResult->nullAt(0) = true;
+      return;
+    }
+
+    auto numArgs = argMapping_.size();
+
+    std::vector<exec::LocalDecodedVector> decodedArgs;
+    decodedArgs.reserve(numArgs);
+
+    for (auto i = 0; i < numArgs; ++i) {
+      auto index = argMapping_[i];
+      if (constantStringViews_[i].empty()) {
+        decodedArgs.emplace_back(context, *args[index], rows);
+      } else {
+        // Do not decode constant inputs.
+        decodedArgs.emplace_back(context);
+      }
+    }
+
+    // Calculate the combined size of the result strings.
+    size_t totalResultBytes = 0;
+    rows.applyToSelected([&](auto row) {
+      auto isFirst = true;
+      for (int i = 0; i < numArgs; i++) {
+        auto value = constantStringViews_[i].empty()
+            ? decodedArgs[i]->valueAt<StringView>(row)
+            : constantStringViews_[i];
+        if (!value.empty()) {
+          if (isFirst) {
+            isFirst = false;
+          } else {
+            totalResultBytes += delim_.size();
+          }
+          totalResultBytes += value.size();
+        }
+      }
+    });
+
+    // Allocate a string buffer.
+    auto rawBuffer = flatResult->getRawStringBufferWithSpace(totalResultBytes);
+    size_t offset = 0;
+    rows.applyToSelected([&](int row) {
+      const char* start = rawBuffer + offset;
+      size_t combinedSize = 0;
+      auto isFirst = true;
+      for (int i = 0; i < numArgs; i++) {
+        StringView value;
+        if (constantStringViews_[i].empty()) {
+          value = decodedArgs[i]->valueAt<StringView>(row);
+        } else {
+          value = constantStringViews_[i];
+        }
+        auto size = value.size();
+        if (size > 0) {
+          if (isFirst) {
+            isFirst = false;
+          } else {
+            memcpy(rawBuffer + offset, delim_.data(), delim_.size());
+            offset += delim_.size();
+            combinedSize += delim_.size();
+          }
+          memcpy(rawBuffer + offset, value.data(), size);
+          combinedSize += size;
+          offset += size;
+        }
+      }
+      flatResult->setNoCopy(row, StringView(start, combinedSize));
+    });
+  }
+
+ private:
+  const std::string delim_;
+  bool nullDelim_ = false;
+  std::vector<column_index_t> argMapping_;
+  std::vector<std::string> constantStrings_;
+  std::vector<StringView> constantStringViews_;
+};
+
+std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+  // varchar, varchar,.. -> varchar.
+  return {exec::FunctionSignatureBuilder()
```

Comment:
Sorry for the late rely. We talk about this way and worried about performance degradation because of type cast.
So we support two most frequently used scenarios :

- concat multi `StringType` by connector
- concat one single `ArrayType(StringType)` by connector